src/coprocessor/endpoint.rs

```diff
@@ -428,69 +377,127 @@ impl Drop for RequestTracker {
     }
 }
 
-#[derive(Debug)]
 pub struct RequestTask {
-    req: Request,
-    cop_req: CopRequest,
-    ctx: ReqContext,
+    req_ctx: Arc<ReqContext>,
+    req_handler_builder: RequestHandlerBuilder,
     on_resp: OnResponse<Response>,
     tracker: RequestTracker,
 }
 
+impl Debug for RequestTask {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("RequestTask")
+            .field("req_ctx", &self.req_ctx)
+            .field("tracker", &self.tracker)
+            .finish()
+    }
+}
+
+impl Display for RequestTask {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Debug::fmt(self, f)
+    }
+}
+
 impl RequestTask {
     pub fn new(
         peer: String,
-        req: Request,
+        mut req: Request,
         on_resp: OnResponse<Response>,
         recursion_limit: u32,
+        batch_row_limit: usize,
+        max_handle_duration: Duration,
     ) -> Result<RequestTask> {
-        let mut table_scan = false;
+        let (data, ranges, context, tp) = (
+            req.take_data(),
+            req.take_ranges(),
+            req.take_context(),
+            req.get_tp(),
+        );
+        // drop it in case of mistakenly using its fields
+        drop(req);
+
+        let mut is_table_scan = false;
         let mut is_desc_scan: Option<bool> = None; // only used in slow query logs
-        let (start_ts, cop_req) = match req.get_tp() {
+        let (req_ctx, req_handler_builder): (_, RequestHandlerBuilder) = match tp {
             REQ_TYPE_DAG => {
-                let mut is = CodedInputStream::from_bytes(req.get_data());
+                let mut is = CodedInputStream::from_bytes(&data);
```

Comment:
Seems L424-L425 repeated in each case, could we move it before `match` L422?

Reply:
nice catch!